File: engines/src/main/scala/olditemrec/ItemRecPreparator.scala
/** Copyright 2014 TappingStone, Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *     http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */

package io.prediction.engines.olditemrec

import io.prediction.controller._

// We reuse the sample business logic as in ItemRankPreparator. Will refactor
// this code into a common util package.

import io.prediction.engines.itemrank.PreparatorParams
import io.prediction.engines.itemrank.ItemRankPreparator
import io.prediction.engines.base.{ PreparedData => IRPreparedData }
import io.prediction.engines.base.{ TrainingData => IRTrainingData }

import org.apache.mahout.cf.taste.model.DataModel
import io.prediction.engines.util.MahoutUtil;

import io.prediction.engines.java.olditemrec.data.PreparedData


class NewItemRecPreparator(pp: PreparatorParams)
  extends LPreparator[
      PreparatorParams,
      IRTrainingData,
      PreparedData] {

  val irPreparator = new ItemRankPreparator(pp)

  override def prepare(irTrainingData: IRTrainingData): PreparedData = {
    val irPreparedData: IRPreparedData = irPreparator.prepare(irTrainingData)

    val ratings: Seq[(Int, Int, Float, Long)] = irPreparedData.rating
    .map { r => (r.uindex, r.iindex, r.rating.toFloat, r.t) }

    val dataModel = MahoutUtil.buildDataModel(ratings)
    new PreparedData(dataModel)
  }

}

File: engines/src/main/java/olditemrec/ItemRecPreparator.java
package io.prediction.engines.java.olditemrec;

import io.prediction.engines.java.olditemrec.data.PreparedData;
import io.prediction.engines.java.olditemrec.data.TrainingData;
import io.prediction.engines.java.olditemrec.data.Rating;
import io.prediction.controller.java.LJavaPreparator;
import io.prediction.controller.java.EmptyParams;
import io.prediction.engines.util.MahoutUtil;

import org.apache.mahout.cf.taste.model.DataModel;
import org.apache.mahout.cf.taste.impl.model.GenericDataModel;
import org.apache.mahout.cf.taste.impl.model.file.FileDataModel;

import scala.Tuple4;
import java.util.List;
import java.util.ArrayList;

/** Preparetor for mahout algo */
public class ItemRecPreparator extends LJavaPreparator<
  EmptyParams, TrainingData, PreparedData> {

  @Override
  public PreparedData prepare(TrainingData td) {
    List<Tuple4<Integer, Integer, Float, Long>> ratings = new ArrayList<
      Tuple4<Integer, Integer, Float, Long>>();
    for (Rating r : td.ratings) {
      ratings.add(new Tuple4<Integer, Integer, Float, Long>(r.uid, r.iid, r.rating, r.t));
    }
    DataModel dataModel = MahoutUtil.jBuildDataModel(ratings);
    PreparedData pd = new PreparedData(dataModel);
    return pd;
  }
}
